openstudiocore/src/geographic_lib/man/GeoidEval.1
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.29)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GEOIDEVAL 1"
.TH GEOIDEVAL 1 "2015-09-30" "GeographicLib 1.45" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
GeoidEval \-\- look up geoid heights
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBGeoidEval\fR [ \fB\-n\fR \fIname\fR ] [ \fB\-d\fR \fIdir\fR ] [ \fB\-l\fR ]
[ \fB\-a\fR | \fB\-c\fR \fIsouth\fR \fIwest\fR \fInorth\fR \fIeast\fR ] [ \fB\-w\fR ]
[ \fB\-z\fR \fIzone\fR ] [ \fB\-\-msltohae\fR ] [ \fB\-\-haetomsl\fR ]
[ \fB\-v\fR ]
[ \fB\-\-comment\-delimiter\fR \fIcommentdelim\fR ]
[ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
[ \fB\-\-input\-file\fR \fIinfile\fR | \fB\-\-input\-string\fR \fIinstring\fR ]
[ \fB\-\-line\-separator\fR \fIlinesep\fR ]
[ \fB\-\-output\-file\fR \fIoutfile\fR ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBGeoidEval\fR reads in positions on standard input and prints out the
corresponding heights of the geoid above the \s-1WGS84\s0 ellipsoid on standard
output.
.PP
Positions are given as latitude and longitude, \s-1UTM/UPS,\s0 or \s-1MGRS,\s0 in any
of the formats accepted by \fIGeoConvert\fR\|(1).  (\s-1MGRS\s0 coordinates signify the
\&\fIcenter\fR of the corresponding \s-1MGRS\s0 square.)  If the \fB\-z\fR option is
specified then the specified zone is prepended to each line of input
(which must be in \s-1UTM/UPS\s0 coordinates).  This allows a file with \s-1UTM\s0
eastings and northings in a single zone to be used as standard input.
.PP
More accurate results for the geoid height are provided by \fIGravity\fR\|(1).
This utility can also compute the direction of gravity accurately.
.PP
The height of the geoid above the ellipsoid, \fIN\fR, is sometimes called
the geoid undulation.  It can be used to convert a height above the
ellipsoid, \fIh\fR, to the corresponding height above the geoid (the
orthometric height, roughly the height above mean sea level), \fIH\fR,
using the relations
.Sp
.RS 4
\&\fIh\fR = \fIN\fR + \fIH\fR, \ \ \fIH\fR = \-\fIN\fR + \fIh\fR.
.RE
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-n\fR" 4
.IX Item "-n"
use geoid \fIname\fR instead of the default \f(CW\*(C`egm96\-5\*(C'\fR.  See
\&\*(L"\s-1GEOIDS\*(R"\s0.
.IP "\fB\-d\fR" 4
.IX Item "-d"
read geoid data from \fIdir\fR instead of the default.  See
\&\*(L"\s-1GEOIDS\*(R"\s0.
.IP "\fB\-l\fR" 4
.IX Item "-l"
use bilinear interpolation instead of cubic.  See
\&\*(L"\s-1INTERPOLATION\*(R"\s0.
.IP "\fB\-a\fR" 4
.IX Item "-a"
cache the entire data set in memory.  See \*(L"\s-1CACHE\*(R"\s0.
.IP "\fB\-c\fR" 4
.IX Item "-c"
cache the data bounded by \fIsouth\fR \fIwest\fR \fInorth\fR \fIeast\fR in memory.
The first two arguments specify the \s-1SW\s0 corner of the cache and the last
two arguments specify the \s-1NE\s0 corner.  The \fB\-w\fR flag specifies that
longitude precedes latitude for these corners, provided that it appears
before \fB\-c\fR.  See \*(L"\s-1CACHE\*(R"\s0.
.IP "\fB\-w\fR" 4
.IX Item "-w"
when reading geographic coordinates, longitude precedes latitude (this
can be overridden by a hemisphere designator, \fIN\fR, \fIS\fR, \fIE\fR, \fIW\fR).
.IP "\fB\-z\fR" 4
.IX Item "-z"
prefix each line of input by \fIzone\fR, e.g., \f(CW\*(C`38n\*(C'\fR.  This should be used
when the input consists of \s-1UTM/UPS\s0 eastings and northings.
.IP "\fB\-\-msltohae\fR" 4
.IX Item "--msltohae"
standard input should include a final token on each line which is
treated as a height (in meters) above the geoid and the output echoes
the input line with the height converted to height above ellipsoid
(\s-1HAE\s0).  If \fB\-z\fR \fIzone\fR is specified then the \fIthird\fR token is treated
as the height; this makes it possible to convert \s-1LIDAR\s0 data where each
line consists of: easting northing height intensity.
.IP "\fB\-\-haetomsl\fR" 4
.IX Item "--haetomsl"
this is similar to \fB\-\-msltohae\fR except that the height token is treated
as a height (in meters) above the ellipsoid and the output echoes the
input line with the height converted to height above the geoid (\s-1MSL\s0).
.IP "\fB\-v\fR" 4
.IX Item "-v"
print information about the geoid on standard error before processing
the input.
.IP "\fB\-\-comment\-delimiter\fR" 4
.IX Item "--comment-delimiter"
set the comment delimiter to \fIcommentdelim\fR (e.g., \*(L"#\*(R" or \*(L"//\*(R").  If
set, the input lines will be scanned for this delimiter and, if found,
the delimiter and the rest of the line will be removed prior to
processing and subsequently appended to the output line (separated by a
space).
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
print version and exit.
.IP "\fB\-h\fR" 4
.IX Item "-h"
print usage, the default geoid path and name, and exit.
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
print full documentation and exit.
.IP "\fB\-\-input\-file\fR" 4
.IX Item "--input-file"
read input from the file \fIinfile\fR instead of from standard input; a file
name of \*(L"\-\*(R" stands for standard input.
.IP "\fB\-\-input\-string\fR" 4
.IX Item "--input-string"
read input from the string \fIinstring\fR instead of from standard input.
All occurrences of the line separator character (default is a semicolon)
in \fIinstring\fR are converted to newlines before the reading begins.
.IP "\fB\-\-line\-separator\fR" 4
.IX Item "--line-separator"
set the line separator character to \fIlinesep\fR.  By default this is a
semicolon.
.IP "\fB\-\-output\-file\fR" 4
.IX Item "--output-file"
write output to the file \fIoutfile\fR instead of to standard output; a
file name of \*(L"\-\*(R" stands for standard output.
.SH "GEOIDS"
.IX Header "GEOIDS"
\&\fBGeoidEval\fR computes geoid heights by interpolating on the data in a
regularly spaced table (see \*(L"\s-1INTERPOLATION\*(R"\s0).  The following geoid
tables are available (however, some may not be installed):
.PP
.Vb 9
\&                                  bilinear error    cubic error
\&   name         geoid    grid     max      rms      max      rms
\&   egm84\-30     EGM84    30\*(Aq      1.546 m  70 mm    0.274 m  14 mm
\&   egm84\-15     EGM84    15\*(Aq      0.413 m  18 mm    0.021 m  1.2 mm
\&   egm96\-15     EGM96    15\*(Aq      1.152 m  40 mm    0.169 m  7.0 mm
\&   egm96\-5      EGM96     5\*(Aq      0.140 m  4.6 mm   .0032 m  0.7 mm
\&   egm2008\-5    EGM2008   5\*(Aq      0.478 m  12 mm    0.294 m  4.5 mm
\&   egm2008\-2_5  EGM2008   2.5\*(Aq    0.135 m  3.2 mm   0.031 m  0.8 mm
\&   egm2008\-1    EGM2008   1\*(Aq      0.025 m  0.8 mm   .0022 m  0.7 mm
.Ve
.PP
By default, the \f(CW\*(C`egm96\-5\*(C'\fR geoid is used.  This may changed by setting
the environment variable \f(CW\*(C`GEOGRAPHICLIB_GEOID_NAME\*(C'\fR or with the \fB\-n\fR
option.  The errors listed here are estimates of the quantization and
interpolation errors in the reported heights compared to the specified
geoid.
.PP
The geoid data will be loaded from a directory specified at compile
time.  This may changed by setting the environment variables
\&\f(CW\*(C`GEOGRAPHICLIB_GEOID_PATH\*(C'\fR or \f(CW\*(C`GEOGRAPHICLIB_DATA\*(C'\fR, or with the \fB\-d\fR
option.  The \fB\-h\fR option prints the default geoid path and name.  Use
the \fB\-v\fR option to ascertain the full path name of the data file.
.PP
Instructions for downloading and installing geoid data are available at
<http://geographiclib.sf.net/html/geoid.html#geoidinst>.
.PP
\&\fB\s-1NOTE\s0\fR: all the geoids above apply to the \s-1WGS84\s0 ellipsoid (\fIa\fR =
6378137 m, \fIf\fR = 1/298.257223563) only.
.SH "INTERPOLATION"
.IX Header "INTERPOLATION"
Cubic interpolation is used to compute the geoid height unless \fB\-l\fR is
specified in which case bilinear interpolation is used.  The cubic
interpolation is based on a least-squares fit of a cubic polynomial to a
12\-point stencil
.PP
.Vb 4
\&   . 1 1 .
\&   1 2 2 1
\&   1 2 2 1
\&   . 1 1 .
.Ve
.PP
The cubic is constrained to be independent of longitude when evaluating
the height at one of the poles.  Cubic interpolation is considerably
more accurate than bilinear; however it results in small discontinuities
in the returned height on cell boundaries.
.SH "CACHE"
.IX Header "CACHE"
By default, the data file is randomly read to compute the geoid heights
at the input positions.  Usually this is sufficient for interactive use.
If many heights are to be computed, use \fB\-c\fR \fIsouth\fR \fIwest\fR \fInorth\fR
\&\fIeast\fR to notify \fBGeoidEval\fR to read a rectangle of data into memory;
heights within the this rectangle can then be computed without any disk
access.  If \fB\-a\fR is specified all the geoid data is read; in the case
of \f(CW\*(C`egm2008\-1\*(C'\fR, this requires about 0.5 \s-1GB\s0 of \s-1RAM. \s0 The evaluation of
heights outside the cached area causes the necessary data to be read
from disk.  Use the \fB\-v\fR option to verify the size of the cache.
.PP
Regardless of whether any cache is requested (with the \fB\-a\fR or \fB\-c\fR
options), the data for the last grid cell in cached.  This allows
the geoid height along a continuous path to be returned with little
disk overhead.
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
.IP "\fB\s-1GEOGRAPHICLIB_GEOID_NAME\s0\fR" 4
.IX Item "GEOGRAPHICLIB_GEOID_NAME"
Override the compile-time default geoid name of \f(CW\*(C`egm96\-5\*(C'\fR.  The \fB\-h\fR
option reports the value of \fB\s-1GEOGRAPHICLIB_GEOID_NAME\s0\fR, if defined,
otherwise it reports the compile-time value.  If the \fB\-n\fR \fIname\fR
option is used, then \fIname\fR takes precedence.

CHANGELOG.md
## OpenStudio v2.2.2
###### 2017-09-13 — [Diff]( https://github.com/NREL/OpenStudio/compare/v2.2.1...v2.2.2 )

- ![Fixed:][fixed] [#2562]( https://github.com/NREL/OpenStudio/issues/2562 ), Apply Measure Now Clears Measures Tab
- ![Fixed:][fixed] [#2602]( https://github.com/NREL/OpenStudio/issues/2602 ), Can't save PAT 2.1.1 project on exit
- ![Fixed:][fixed] [#2730]( https://github.com/NREL/OpenStudio/issues/2730 ), Opening existing floorplan in editor freezes app
- ![Fixed:][fixed] [#2763]( https://github.com/NREL/OpenStudio/issues/2763 ), PAT Doesn't Show Analysis Complete for Modest Cloud Runs

## OpenStudio v2.2.1
###### 2017-09-06 — [Diff]( https://github.com/NREL/OpenStudio/compare/v2.2.0...v2.2.1 )

- ![Fixed:][fixed] [#162]( https://github.com/NREL/OpenStudio/issues/162 ), Typo in People Definition tab (Bugzilla #840)
- ![Fixed:][fixed] [#1632]( https://github.com/NREL/OpenStudio/issues/1632 ), Incorrect results in OpenStudio App.
- ![Fixed:][fixed] [#2317]( https://github.com/NREL/OpenStudio/issues/2317 ), incorrct areas in EnergyPlus Results
- ![Fixed:][fixed] [#2397]( https://github.com/NREL/OpenStudio/issues/2397 ),  OpenStudio SketchUp 2016 crashed when opens
- ![Fixed:][fixed] [#2457]( https://github.com/NREL/OpenStudio/issues/2457 ), OS App Units missing on Materials and Internal Loads (except for people) in 2.0.1
- ![Improved:][improved] [#2500]( https://github.com/NREL/OpenStudio/issues/2500 ), Make PAT "Additional Analysis Files" path relative
- ![Fixed:][fixed] [#2619]( https://github.com/NREL/OpenStudio/issues/2619 ), Successive add/delete of Measures in App's Measures Tab Creates Cruft
- ![Fixed:][fixed] [#2629]( https://github.com/NREL/OpenStudio/issues/2629 ), Duplicated measure results are overwritten
- ![Improved:][improved] [#2632]( https://github.com/NREL/OpenStudio/issues/2632 ), Server takes to long to start
- ![Fixed:][fixed] [#2669]( https://github.com/NREL/OpenStudio/issues/2669 ), Prevent user from naming PAT project 'PAT'
- ![Fixed:][fixed] [#2676]( https://github.com/NREL/OpenStudio/issues/2676 ), PAT 2.1.2 deletes measure from library
- ![Fixed:][fixed] [#2677]( https://github.com/NREL/OpenStudio/issues/2677 ), PAT Results Summary Table Report has wrong sign for capital cost increase
- ![Fixed:][fixed] [#2681]( https://github.com/NREL/OpenStudio/issues/2681 ), Surface disappears after saving in SketchUp - after using Intersect and Divide Inter-Zone Surface for the entire model
- ![Fixed:][fixed] [#2688]( https://github.com/NREL/OpenStudio/issues/2688 ), Supply air fan operating mode
- ![Fixed:][fixed] [#2690]( https://github.com/NREL/OpenStudio/issues/2690 ), Duplicate Measure, breaks if you then update the measure
- ![Fixed:][fixed] [#2691]( https://github.com/NREL/OpenStudio/issues/2691 ), About PAT window hasn't been updated in agest
- ![Fixed:][fixed] [#2694]( https://github.com/NREL/OpenStudio/issues/2694 ), Segfault in Materials Tab
- ![Fixed:][fixed] [#2695]( https://github.com/NREL/OpenStudio/issues/2695 ), User Reports PAT Local Server Not Starting Reliably on Mac
- ![Improved:][improved] [#2716]( https://github.com/NREL/OpenStudio/issues/2716 ), AWS deployed server needs to run 350k data-points
- ![Fixed:][fixed] [#2719]( https://github.com/NREL/OpenStudio/issues/2719 ), PAT hangs when running simulation on the cloud
- ![Fixed:][fixed] [#2721]( https://github.com/NREL/OpenStudio/issues/2721 ), PAT thinks cluster is running (but it's not)
- ![Fixed:][fixed] [#2723]( https://github.com/NREL/OpenStudio/issues/2723 ), Third PAT algorithm argument in Morris should be Grid Jump
- ![Fixed:][fixed] [#2724]( https://github.com/NREL/OpenStudio/issues/2724 ), Morris Method Box Plots are Empty in 2.2.0
- ![Fixed:][fixed] [#2728]( https://github.com/NREL/OpenStudio/issues/2728 ), PAT Analysis Error when updating measure w/ duplicates
- ![Fixed:][fixed] [#2729]( https://github.com/NREL/OpenStudio/issues/2729 ), Place Map dialog no longer pops up on new editor
- ![Fixed:][fixed] [#2732]( https://github.com/NREL/OpenStudio/issues/2732 ), No way to dismiss preview window
- ![Fixed:][fixed] [#2733]( https://github.com/NREL/OpenStudio/issues/2733 ), Geometry boundary conditions in editor preview are wrong
- ![Improved:][improved] [#2734]( https://github.com/NREL/OpenStudio/issues/2734 ), No option to continue editing geometry after merge operation
- ![Fixed:][fixed] [#2736]( https://github.com/NREL/OpenStudio/issues/2736 ), Returning to geometry editor pane after merge produces inconsistent results
- ![Improved:][improved] [#2737]( https://github.com/NREL/OpenStudio/issues/2737 ), No feedback from App post-merge
- ![Fixed:][fixed] [#2738]( https://github.com/NREL/OpenStudio/issues/2738 ), Boundary condition between floors doesn't appear correct
- ![Fixed:][fixed] [#2740]( https://github.com/NREL/OpenStudio/issues/2740 ), Floor height isn't translated
- ![Fixed:][fixed] [#2745]( https://github.com/NREL/OpenStudio/issues/2745 ), Going to Preview Tab throws away unmerged changes
- ![Fixed:][fixed] [#2746]( https://github.com/NREL/OpenStudio/issues/2746 ), Prevent PAT from running projects with zero variables
- ![Fixed:][fixed] [#2748]( https://github.com/NREL/OpenStudio/issues/2748 ), Levels in Morris Method Must be >1
- ![Fixed:][fixed] [#2754]( https://github.com/NREL/OpenStudio/issues/2754 ), App locks up on image import
- ![Fixed:][fixed] [#2760]( https://github.com/NREL/OpenStudio/issues/2760 ), PAT Not removing Measures from Projects

## OpenStudio v2.2.0
###### 2017-06-30 — [Diff]( https://github.com/NREL/OpenStudio/compare/v2.1.2...v2.2.0 )

- ![Fixed:][fixed] [#2496]( https://github.com/NREL/OpenStudio/issues/2496 ), Chinese characters in Design Alternatives grid
- ![Fixed:][fixed] [#2503]( https://github.com/NREL/OpenStudio/issues/2503 ), Skipping measures in datapoint throws measure messages out of sync

openstudiocore/src/geographic_lib/tools/GeoidEval.cpp
/**
 * \file GeoidEval.cpp
 * \brief Command line utility for evaluating geoid heights
 *
 * Copyright (c) Charles Karney (2009-2015) <charles@karney.com> and licensed
 * under the MIT/X11 License.  For more information, see
 * http://geographiclib.sourceforge.net/
 *
 * See the <a href="GeoidEval.1.html">man page</a> for usage information.
 **********************************************************************/

#include <iostream>
#include <string>
#include <sstream>
#include <fstream>
#include <GeographicLib/Geoid.hpp>
#include <GeographicLib/DMS.hpp>
#include <GeographicLib/Utility.hpp>
#include <GeographicLib/GeoCoords.hpp>

#if defined(_MSC_VER)
// Squelch warnings about constant conditional expressions and potentially
// uninitialized local variables
#  pragma warning (disable: 4127 4701)
#endif

#include "GeoidEval.usage"

int main(int argc, char* argv[]) {
  try {
    using namespace GeographicLib;
    typedef Math::real real;
    Utility::set_digits();
    bool cacheall = false, cachearea = false, verbose = false,
      cubic = true, gradp = false;
    real caches, cachew, cachen, cachee;
    std::string dir;
    std::string geoid = Geoid::DefaultGeoidName();
    Geoid::convertflag heightmult = Geoid::NONE;
    std::string istring, ifile, ofile, cdelim;
    char lsep = ';';
    bool northp = false, longfirst = false;
    int zonenum = UTMUPS::INVALID;

    for (int m = 1; m < argc; ++m) {
      std::string arg(argv[m]);
      if (arg == "-a") {
        cacheall = true;
        cachearea = false;
      }
      else if (arg == "-c") {
        if (m + 4 >= argc) return usage(1, true);
        cacheall = false;
        cachearea = true;
        try {
          DMS::DecodeLatLon(std::string(argv[m + 1]), std::string(argv[m + 2]),
                            caches, cachew, longfirst);
          DMS::DecodeLatLon(std::string(argv[m + 3]), std::string(argv[m + 4]),
                            cachen, cachee, longfirst);
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding argument of -c: " << e.what() << "\n";
          return 1;
        }
        m += 4;
      } else if (arg == "--msltohae")
        heightmult = Geoid::GEOIDTOELLIPSOID;
      else if (arg == "--haetomsl")
        heightmult = Geoid::ELLIPSOIDTOGEOID;
      else if (arg == "-w")
        longfirst = true;
      else if (arg == "-z") {
        if (++m == argc) return usage(1, true);
        std::string zone = argv[m];
        try {
          UTMUPS::DecodeZone(zone, zonenum, northp);
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding zone: " << e.what() << "\n";
          return 1;
        }
        if (!(zonenum >= UTMUPS::MINZONE && zonenum <= UTMUPS::MAXZONE)) {
          std::cerr << "Illegal zone " << zone << "\n";
          return 1;
        }
      } else if (arg == "-n") {
        if (++m == argc) return usage(1, true);
        geoid = argv[m];
      } else if (arg == "-d") {
        if (++m == argc) return usage(1, true);
        dir = argv[m];
      } else if (arg == "-l")
        cubic = false;
      else if (arg == "-g")
        gradp = true;
      else if (arg == "-v")
        verbose = true;
      else if (arg == "--input-string") {
        if (++m == argc) return usage(1, true);
        istring = argv[m];
      } else if (arg == "--input-file") {
        if (++m == argc) return usage(1, true);
        ifile = argv[m];
      } else if (arg == "--output-file") {
        if (++m == argc) return usage(1, true);
        ofile = argv[m];
      } else if (arg == "--line-separator") {
        if (++m == argc) return usage(1, true);
        if (std::string(argv[m]).size() != 1) {
          std::cerr << "Line separator must be a single character\n";
          return 1;
        }
        lsep = argv[m][0];
      } else if (arg == "--comment-delimiter") {
        if (++m == argc) return usage(1, true);
        cdelim = argv[m];
      } else if (arg == "--version") {
        std::cout << argv[0] << ": GeographicLib version "
                  << GEOGRAPHICLIB_VERSION_STRING << "\n";
        return 0;
      } else {
        int retval = usage(!(arg == "-h" || arg == "--help"), arg != "--help");
        if (arg == "-h")
          std::cout<< "\nDefault geoid path = \""   << Geoid::DefaultGeoidPath()
                   << "\"\nDefault geoid name = \"" << Geoid::DefaultGeoidName()
                   << "\"\n";
        return retval;
      }
    }

    if (!ifile.empty() && !istring.empty()) {
      std::cerr << "Cannot specify --input-string and --input-file together\n";
      return 1;
    }
    if (ifile == "-") ifile.clear();
    std::ifstream infile;
    std::istringstream instring;
    if (!ifile.empty()) {
      infile.open(ifile.c_str());
      if (!infile.is_open()) {
        std::cerr << "Cannot open " << ifile << " for reading\n";
        return 1;
      }
    } else if (!istring.empty()) {
      std::string::size_type m = 0;
      while (true) {
        m = istring.find(lsep, m);
        if (m == std::string::npos)
          break;
        istring[m] = '\n';
      }
      instring.str(istring);
    }
    std::istream* input = !ifile.empty() ? &infile :
      (!istring.empty() ? &instring : &std::cin);

    std::ofstream outfile;
    if (ofile == "-") ofile.clear();
    if (!ofile.empty()) {
      outfile.open(ofile.c_str());
      if (!outfile.is_open()) {
        std::cerr << "Cannot open " << ofile << " for writing\n";
        return 1;
      }
    }
    std::ostream* output = !ofile.empty() ? &outfile : &std::cout;

    int retval = 0;
    try {
      const Geoid g(geoid, dir, cubic);
      try {
        if (cacheall)
          g.CacheAll();
        else if (cachearea)
          g.CacheArea(caches, cachew, cachen, cachee);
      }
      catch (const std::exception& e) {
        std::cerr << "ERROR: " << e.what() << "\nProceeding without a cache\n";
      }
      if (verbose) {
        std::cerr << "Geoid file: "    << g.GeoidFile()     << "\n"
                  << "Description: "   << g.Description()   << "\n"
                  << "Interpolation: " << g.Interpolation() << "\n"
                  << "Date & Time: "   << g.DateTime()      << "\n"
                  << "Offset (m): "    << g.Offset()        << "\n"
                  << "Scale (m): "     << g.Scale()         << "\n"
                  << "Max error (m): " << g.MaxError()      << "\n"
                  << "RMS error (m): " << g.RMSError()      << "\n";
        if (g.Cache())
          std::cerr<< "Caching:"
                   << "\n SW Corner: " << g.CacheSouth() << " " << g.CacheWest()
                   << "\n NE Corner: " << g.CacheNorth() << " " << g.CacheEast()
                   << "\n";
      }

      GeoCoords p;
      std::string s, eol, suff;
      const char* spaces = " \t\n\v\f\r,"; // Include comma as space
      while (std::getline(*input, s)) {
        try {
          eol = "\n";
          if (!cdelim.empty()) {
            std::string::size_type m = s.find(cdelim);
            if (m != std::string::npos) {
              eol = " " + s.substr(m) + "\n";
              std::string::size_type m1 =
                m > 0 ? s.find_last_not_of(spaces, m - 1) : std::string::npos;
              s = s.substr(0, m1 != std::string::npos ? m1 + 1 : m);
            }
          }
          real height = 0;
          if (zonenum != UTMUPS::INVALID) {
            // Expect "easting northing" if heightmult == 0, or
            // "easting northing height" if heightmult != 0.
            std::string::size_type pa = 0, pb = 0;
            real easting = 0, northing = 0;
            for (int i = 0; i < (heightmult ? 3 : 2); ++i) {
              if (pb == std::string::npos)
                throw GeographicErr("Incomplete input: " + s);
              // Start of i'th token
              pa = s.find_first_not_of(spaces, pb);
              if (pa == std::string::npos)
                throw GeographicErr("Incomplete input: " + s);
              // End of i'th token
              pb = s.find_first_of(spaces, pa);
              (i == 2 ? height : (i == 0 ? easting : northing)) =
                Utility::num<real>(s.substr(pa, (pb == std::string::npos ?
                                                 pb : pb - pa)));
            }
            p.Reset(zonenum, northp, easting, northing);
            if (heightmult) {
              suff = pb == std::string::npos ? "" : s.substr(pb);
              s = s.substr(0, pa);
            }
          } else {
            if (heightmult) {
              // Treat last token as height
              // pb = last char of last token
              // pa = last char preceding white space
              // px = last char of 2nd last token
              std::string::size_type pb = s.find_last_not_of(spaces);
              std::string::size_type pa = s.find_last_of(spaces, pb);
              if (pa == std::string::npos || pb == std::string::npos)
                throw GeographicErr("Incomplete input: " + s);
              height = Utility::num<real>(s.substr(pa + 1, pb - pa));
              s = s.substr(0, pa + 1);
            }
            p.Reset(s, true, longfirst);
          }
          if (heightmult) {
            real h = g(p.Latitude(), p.Longitude());
            *output << s
                    << Utility::str(height + real(heightmult) * h, 4)
                    << suff << eol;
          } else {
            if (gradp) {
            real gradn, grade;
            real h = g(p.Latitude(), p.Longitude(), gradn, grade);
            *output << Utility::str(h, 4) << " "
                    << Utility::str(gradn * 1e6, 2)
                    << (Math::isnan(gradn) ? " " : "e-6 ")
                    << Utility::str(grade * 1e6, 2)
                    << (Math::isnan(grade) ? "" : "e-6")
                    << eol;
            } else {
            real h = g(p.Latitude(), p.Longitude());
            *output << Utility::str(h, 4) << eol;
            }
          }
        }
        catch (const std::exception& e) {
          *output << "ERROR: " << e.what() << "\n";
          retval = 1;
        }
      }
    }
    catch (const std::exception& e) {
      std::cerr << "Error reading " << geoid << ": " << e.what() << "\n";
      retval = 1;
    }
    return retval;
  }
  catch (const std::exception& e) {
    std::cerr << "Caught exception: " << e.what() << "\n";
    return 1;
  }
  catch (...) {
    std::cerr << "Caught unknown exception\n";
    return 1;
  }
}
